Add support for EC2 instance selector

cPu1 opened the pull request:
### Description

Adds support for EC2 instance selector. 

Closes https://github.com/weaveworks/eksctl/issues/3461

TODO:
- [x] This is based off the `dry-run` branch which isn't merged yet. Wait for it to be merged
- [x] Add unit tests
- [x] Add integration tests

### Checklist
- [x] Added tests that cover your change (if possible)
- [ ] Added/modified documentation as required (such as the `README.md`, or the `userdocs` directory)
- [x] Manually tested
- [x] Added labels for change area (e.g. `area/nodegroup`), target version (e.g. `version/0.12.0`) and kind (e.g. `kind/improvement`)
- [x] Make sure the title of the PR is a good description that can go into the release notes

Callisto13 reviewed (changes requested):
🎉 . Didn't know the selector had a library! Very happy about that 😂 

Just needs tests added to:
- pkg/apis/eksctl.io/v1alpha5/validation_test.go
- pkg/apis/eksctl.io/v1alpha5/defaults_test.go
- pkg/eks/nodegroup_service_test.go

Are you adding an example along with the docs?

Callisto13 commented on pkg/apis/eksctl.io/v1alpha5/types.go:
```diff
@@ -1456,6 +1460,27 @@ type PrivateCluster struct {
+type InstanceSelector struct {
+	// VCPUs specifies the number of vCPUs
+	VCPUs int `json:"vCPUs,omitempty"`
+	// Memory specifies the memory in GiB
+	Memory string `json:"memory,omitempty"`
```
does the memory type need to be a `string` if the unit is predetermined? i can see users assuming it is `int` like the others and getting an error

cPu1 replied:
That comment is not accurate, the unit defaults to GiB but can be set to MiB etc. I'll update the comment.

Callisto13 replied:
aah ok makes sense 👍

Callisto13 commented on pkg/apis/eksctl.io/v1alpha5/types.go:
```diff
@@ -1456,6 +1460,27 @@ type PrivateCluster struct {
+// InstanceSelector holds EC2 instance selector options
+type InstanceSelector struct {
+	// VCPUs specifies the number of vCPUs
+	VCPUs int `json:"vCPUs,omitempty"`
```
the casing of `vCPUs` compared to `gpus` is a bit weird

cPu1 replied:
Yeah, I had highlighted this in the proposal: https://github.com/weaveworks/eksctl/blob/main/docs/proposal-007-instance-selector-dry-run.md#design

>Note: Because fields are case-sensitive, vCPUs may be difficult to get right, especially for users unfamiliar with Go naming conventions. However, vCPUs with the same case is standard enough that it's widely used by all major cloud providers. vcpus is easier to remember but it's inconsistent with the other field names.

cPu1 commented:
> Just needs tests added to:

> * pkg/apis/eksctl.io/v1alpha5/validation_test.go

I have added validation tests to `pkg/apis/eksctl.io/v1alpha5/instance_selector_test.go` and `pkg/apis/eksctl.io/v1alpha5/managed_nodegroup_test.go`

> * pkg/apis/eksctl.io/v1alpha5/defaults_test.go

The defaulting part is covered by `pkg/ctl/cmdutils/filter/nodegroup_filter_test.go`.

> * pkg/eks/nodegroup_service_test.go

I added an integration test that covers what this package does.





> Are you adding an example along with the docs?

Yes! But I'm now planning to add them in a separate PR so we aren't blocked on merging this.

Callisto13 commented on integration/tests/dry_run/dry_run_test.go:
```diff
@@ -216,7 +223,59 @@ var _ = Describe("(Integration) [Dry-Run test]", func() {
+		}, "--managed", "--instance-selector-vcpus=2", "--instance-selector-memory=4", "--instance-selector-cpu-architecture=x86_64"),
+	)
+
+	PDescribe("create cluster and nodegroups from the output of dry-run", func() {
```
should this be Pending?

cPu1 replied:
Good catch, I forgot to update it. Thanks!

Callisto13 commented:
> > pkg/eks/nodegroup_service_test.go

> I added an integration test that covers what this package does.

🤔  I don't know if we want to rely on integration tests for testing core logic. it's quite a long way to go for a failure when adding more code, plus we only do the happy path. there is a decent amount of logic in there and several places where it can error out

Callisto13 commented:
> > Are you adding an example along with the docs?

> Yes! But I'm now planning to add them in a separate PR so we aren't blocked on merging this.

👍  makes sense will we wait for that docs pr to release?
